onnxruntime/test/contrib_ops/ngram_test.cc

```diff
@@ -0,0 +1,410 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "gtest/gtest.h"
+#include "test/providers/provider_test_utils.h"
+
+#include <stdint.h>
+
+namespace onnxruntime {
+namespace test {
+namespace ngram_test {
+
+constexpr const char* domain = onnxruntime::kMSDomain;
+const int opset_ver = 1;
+
+void InitTestAttr(OpTester& test, const std::string& mode,
+                  int64_t M, int64_t N, int64_t S,
+                  const std::vector<int64_t>& ngram_counts,
+                  const std::vector<int64_t>& ngram_indexes,
+                  const std::vector<float>& weights,
+                  const std::vector<int64_t>& pool_int64s,
+                  const std::vector<std::string>& pool_strings) {
+  test.AddAttribute("mode", mode);
+  test.AddAttribute("M", M);
+  test.AddAttribute("N", N);
+  test.AddAttribute("S", S);
+  test.AddAttribute("ngram_counts", ngram_counts);
+  test.AddAttribute("ngram_indexes", ngram_indexes);
+  // optional
+  if (!weights.empty()) {
+    test.AddAttribute("weights", weights);
+  }
+  if (!pool_int64s.empty()) {
+    test.AddAttribute("pool_int64s", pool_int64s);
+  } else {
+    test.AddAttribute("pool_strings", pool_strings);
+  }
+}
+}  // namespace ngram_test
+
+using namespace ngram_test;
+
+// Here is what takes place in general and in particular
+// in this unit test.There are 7 n - grams : 4 unigrams and 3 bigrams
+// that are expressed as 10 items(integers in this case) contained within pool_int64 attribute.
+// We only count and then optionally scale those ngrams that appear in the supplied pool parameter(either int64 or string).
+// M = 1 and N = 2 in this case.
+// However, attribute all controls whether we consider all of the supplied ngram[M..N] sizes
+// into consideration or not.With all = false, we only consider N - grams.
+
+TEST(ContribOpTest, Ngram_Int32_TF_onlyBigrams_Skip0) {
+  OpTester test("Ngram", opset_ver, domain);
+  // 1 - 2, s=0, , M=N, weights empty, int32
+  InitTestAttr(test, "TF", 2, 2, 0,
+               {0, 4},
+               {0, 1, 2, 3, 4, 5, 6},  //7 output indexes
+               {},
+               {2, 3, 5, 4,         //1-grams
+                5, 6, 7, 8, 6, 7},  //bi-grams
+               {});
+
+  std::vector<int64_t> dims{2, 6};
```

Comment:
>dims{2, 6}; [](start = 23, length = 11)

Still not understanding the shape here. If you match '7,8' but the input is of shape {2,6} that would mean the 7 is the end of the first row, and the 8 is the start of the second row. So what does the first dimension represent?

Reply:
The spec has changed (see the other comment) and I will re-implement accordingly.

---
In reply to: [243373007](https://github.com/Microsoft/onnxruntime/pull/180#discussion_r243373007) [](ancestors = 243373007,243148525)